core/src/main/java/org/elasticsearch/script/ScriptMetaData.java

```diff
@@ -18,168 +18,414 @@
+ * as part of the {@link ClusterState}.  Currently scripts can
+ * be stored as part of the new namespace for a stored script where
+ * only an id is used or as part of the deprecated namespace where
+ * both a language and an id are used.
+ */
 public final class ScriptMetaData implements MetaData.Custom {
 
-    public static final String TYPE = "stored_scripts";
-    public static final ScriptMetaData PROTO = new ScriptMetaData(Collections.emptyMap());
+    /**
+     * A builder used to modify the currently stored scripts data held within
+     * the {@link ClusterState}.  Scripts can be added or deleted, then built
+     * to generate a new {@link Map} of scripts that will be used to update
+     * the current {@link ClusterState}.
+     */
+    public static final class Builder {
 
-    private final Map<String, ScriptAsBytes> scripts;
+        private final Map<String, StoredScriptSource> scripts;
 
-    ScriptMetaData(Map<String, ScriptAsBytes> scripts) {
-        this.scripts = scripts;
-    }
+        /**
+         * @param previous The current {@link ScriptMetaData} or {@code null} if there
+         *                 is no existing {@link ScriptMetaData}.
+         */
+        public Builder(ScriptMetaData previous) {
+            this.scripts = previous == null ? new HashMap<>() :new HashMap<>(previous.scripts);
+        }
 
-    public BytesReference getScriptAsBytes(String language, String id) {
-        ScriptAsBytes scriptAsBytes = scripts.get(toKey(language, id));
-        if (scriptAsBytes != null) {
-            return scriptAsBytes.script;
-        } else {
-            return null;
+        /**
+         * Add a new script to the existing stored scripts.  The script will be added under
+         * both the new namespace and the deprecated namespace, so that look ups under
+         * the deprecated namespace will continue to work.  Should a script already exist under
+         * the new namespace using a different language, it will be replaced and a deprecation
+         * warning will be issued.  The replaced script will still exist under the deprecated
+         * namespace and can continue to be looked up this way until it is deleted.
+         * <p>
+         * Take for example script 'A' with lang 'L0' and data 'D0'.  If we add script 'A' to the
+         * empty set, the scripts {@link Map} will be ["A" -- D0, "A#L0" -- D0].  If a script
+         * 'A' with lang 'L1' and data 'D1' is then added, the scripts {@link Map} will be
+         * ["A" -- D1, "A#L1" -- D1, "A#L0" -- D0].
+         * @param id The user-specified id to use for the look up.
+         * @param source The user-specified stored script data held in {@link StoredScriptSource}.
+         */
+        public Builder storeScript(String id, StoredScriptSource source) {
+            StoredScriptSource previous = scripts.put(id, source);
+            scripts.put(source.getLang() + "#" + id, source);
```

Comment:
I thought about this as well and it certainly makes sense, but the tricky part is the user has no choice for stored templates.  I wanted to make sure that the user could still have a stored template with the same name as a script for this version so they have time to come up with an appropriate plan for the deprecation.  With the correct planning the old scheme will simply disappear when loading the scripts from cluster state metadata because we can make it ignore (and print a warning) that an old script cannot be loaded similar to how file scripts work now for scripts that do not compile.

Also, right now this technically isn't a breaking change which I think supports the user a little better (though, hopefully no one is dependent on having scripts of the same name with different languages).